CM7/Core/Src/gatedriver.c

```diff
@@ -21,10 +22,27 @@ static void gatedrv_fault_cb(gatedriver_t* drv)
 
 }
 
-gatedriver_t* gatedrv_init(TIM_HandleTypeDef* tim, ADC_HandleTypeDef *hdma_adc, SPI_HandleTypeDef *adc_spi)
+static void (*gatedrv_hdma_adc_cb_gen(gatedriver_t* drv))(DMA_HandleTypeDef*)
```

Comment:
This feels like a very weird syntax for this. You might be able to just do `void gatedrv_hdma_adc_cb(dma)` or whatever the DMA_HandleTypeDef really is. This feels very confusing to read

Reply:
Wait yeah, you can just define it as `void blahblahblah(DMA_HandleTypeDef *hdma)`. The weird syntax you saw in the register callback function is just the way that you can reference a pointer to a function.

Reply:
That poses the challenge of "how do I retrieve the actual queue from that" which is an equally annoying thing to have to handle because interrupts usually lack a lot of context. Might be able to use some pointer magic to accomplish this like the Linux kernel does

Reply:
An extremely hack way of doing this is to take the pointer, using the register address and the known size of the struct its in, you can calculate the offset of the actual struct and cast that memory to the type of the container. The Linux kernel does this frequently for interrupts using the `container_of` macro. Check it out here:
https://radek.io/2012/11/10/magical-container_of-macro/

Might be extremely janky for our setup and there might be a much better way so def do research, but it is an option

Reply:
Yeah so this weird signature is actually an attempt at getting the `gatedriver` struct during an interrupt. The signature essentially says the function returns a pointer to a callback. I realized the callback doesn't actually get the `gatedriver` struct as a parameter or anything, but it's needed because the buffer for the DMA is in there. In any other language, I would just make a callback with a signature with a different signature and just use lambda to wrap it later, but C doesn't support that. I can poke around to see if there are any other options, though.

Reply:
I'm not entirely sure the `container_of` macro would work here, purely because we only get a pointer to the DMA struct which could really be anywhere in memory. We don't store the actual DMA struct in the `gatedriver_t` struct, so I don't really think we can play those memory games here with complete confidence. I mean my mind went to the function generator approach, but like if you can think of a better solution I'm definitely open to it.